RavenDB-18634 - Use TCP compression for the cluster communication

shaharhikri opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18634/Use-TCP-compression-for-the-cluster-communication

### Additional description

Use TCP compression for the cluster communication.

_Please delete below the options that are not relevant_

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -3620,6 +3620,12 @@ public override async Task<RachisConnection> ConnectToPeer(string url, string ta
 
+                    if (supportedFeatures.DataCompression)
+                    {
+                        stream = new ReadWriteCompressedStream(stream);
```
this is when you create the connection, where is this handled on the receiving end?

shaharhikri replied:
https://github.com/ravendb/ravendb/blob/2b7dd28d2e3a0a4b2d2bfb5b8bd2143ddbc10743/src/Raven.Server/RavenServer.cs#L1958

```
                            if (ShouldUseDataCompression(header))
                            {
                                stream = new ReadWriteCompressedStream(stream, buffer);
                                tcp.Stream = stream;
                            }
```
look at the changes on the ```ShouldUseDataCompression``` method.

karmeli87 commented:
1. I don't see any tests (it is okay to test manually, but note in the PR that it was manually tested)
- We need to test mixed cluster behavior
- Check that disabling the TCP compression doesn't break anything

2. We want to have this also for the `Heartbeats` connection (the maintenance connection between the followers & the leader)

shaharhikri committed:
RavenDB-18634 - Add compression in Heartbeats

shaharhikri commented:
> 1. I don't see any tests (it is okay to test manually, but note in the PR that it was manually tested)
> 
> * We need to test mixed cluster behavior
> * Check that disabling the TCP compression doesn't break anything
> 
> 2. We want to have this also for the `Heartbeats` connection (the maintenance connection between the followers & the leader)

tested mixed cluster (5.4 with cluster/heartbeats TCP compression and 5.2) manually.
added test of a cluster with one node that TCP compression is disabled.
Added compression in heartbeats.

shaharhikri committed:
RavenDB-18634 - Fix Failing tests

ayende commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -43,6 +43,7 @@ public class RachisConsensus<TStateMachine> : RachisConsensus
         where TStateMachine : RachisStateMachine, new()
     {
         private readonly ServerStore _serverStore;
+        //internal ServerStore ServerStore => _serverStore;
```
?

shaharhikri replied:
Removed

karmeli87 commented on src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs:
```diff
@@ -84,8 +85,8 @@ public DynamicJsonValue ToJson()
-        public static readonly int ClusterTcpVersion = ClusterWithMultiTree;
-        public static readonly int HeartbeatsTcpVersion = Heartbeats42000;
+        public static readonly int ClusterTcpVersion = TcpConnectionsWithCompression;
+        public static readonly int HeartbeatsTcpVersion = Heartbeats53000;
```
let's rename it to `HeartbeatsWithTcpCompression`

shaharhikri replied:
Done

karmeli87 commented on src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs:
```diff
@@ -84,8 +85,8 @@ public DynamicJsonValue ToJson()
 
-        public static readonly int ClusterTcpVersion = ClusterWithMultiTree;
-        public static readonly int HeartbeatsTcpVersion = Heartbeats42000;
+        public static readonly int ClusterTcpVersion = TcpConnectionsWithCompression;
```
Have a dedicate one for the the cluster, e.g. `ClusterWithTcpCompression`

shaharhikri replied:
There's no constant with that name.
I added this constant to ```TcpConnectionHeader```.

karmeli87 commented on src/Raven.Server/ServerWide/Maintenance/ClusterMaintenanceSupervisor.cs:
```diff
@@ -467,7 +470,7 @@ private async Task<ClusterMaintenanceConnection> ConnectToClientNodeAsync(TcpCon
                 TcpClient tcpClient;
 
-                using (_contextPool.AllocateOperationContext(out JsonOperationContext ctx))
+                using (var ctx = JsonOperationContext.ShortTermSingleUse())
```
why this change?

shaharhikri replied:
Done

shaharhikri committed:
RavenDB-18634 - Addressed PR Comments

ppekrol commented on src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs:
```diff
@@ -66,9 +66,11 @@ public DynamicJsonValue ToJson()
         public static readonly int HeartbeatsBaseLine = 20;
         public static readonly int Heartbeats41200 = 41_200;
         public static readonly int Heartbeats42000 = 42_000;
+        public static readonly int HeartbeatsWithTcpCompression = 53_000;
```
shouldn't this be 54_000?

shaharhikri replied:
Done

ppekrol commented on src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs:
```diff
@@ -66,9 +66,11 @@ public DynamicJsonValue ToJson()
         public static readonly int DropBaseLine = -2;
         public static readonly int ClusterBaseLine = 10;
         public static readonly int ClusterWithMultiTree = 52_000;
+        public static readonly int ClusterWithTcpCompression = 53_000;
```
shouldn't this be 54_000?

shaharhikri replied:
Done

ppekrol commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -3666,6 +3672,13 @@ await using (stream)
         private async Task<TcpConnectionHeaderMessage.SupportedFeatures> NegotiateProtocolVersionAsyncForCluster(string url, TcpConnectionInfo info, Stream stream, JsonOperationContext ctx, string tag)
         {
+            bool compressionSupport = false;
+#if NETCOREAPP3_1_OR_GREATER
```
can be removed? we are running on .NET 7 so this is alwas true and this is in Raven.Server, not Raven.Client?

shaharhikri replied:
you're right.
it's only relevant to the client, not to the server.
It was removed.

ppekrol reviewed (commented):
Please rebase

shaharhikri committed:
RavenDB-18634 - Use TCP compression for the cluster communication

shaharhikri commented:
Rebased

shaharhikri committed:
RavenDB-18634 - 'HeartbeatsWithTcpCompression' and 'ClusterWithTcpCompression' are now in 5.4 and above

karmeli87 commented on src/Raven.Server/ServerWide/Maintenance/ClusterMaintenanceSupervisor.cs:
```diff
@@ -482,10 +485,12 @@ private async Task<ClusterMaintenanceConnection> ConnectToClientNodeAsync(TcpCon
                     supportedFeatures = result.SupportedFeatures;
 
-                    await using (var writer = new AsyncBlittableJsonTextWriter(ctx, connection))
+                    if (result.SupportedFeatures != null && result.SupportedFeatures.DataCompression)
```
can `SupportedFeatures` be `null`?

shaharhikri replied:
No, I removed it

shaharhikri committed:
RavenDB-18634 - Remove null check of 'SupportedFeatures' in 'ConnectToClientNodeAsync'
